### PR TITLE
[Constraint solver] Sink down initialization pattern handling

### PR DESCRIPTION
**Pull request by DougGregor**
Sink the constraint generation and solution application of initialization patterns, including all
of the logic for property wrappers, from the high-level entry point
`typeCheckBinding` down into the lower-level handling for
solution application targets.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

File: lib/Sema/CSApply.cpp [hunk 1]
```diff
@@ -7267,6 +7267,10 @@ Optional<SolutionApplicationTarget> ConstraintSystem::applySolution(
     if (!rewrittenExpr)
       return None;
 
+    if (target.getExprContextualTypePurpose() == CTP_Initialization) {
+
```
Comment (xedin) [inline lib/Sema/CSApply.cpp]: This seems either incomplete or obsolete...

**Review by xedin — APPROVED** [review]
LGTM!

**Comment by futurejones** [comment]
Can one of the admins verify this patch?

File: lib/Sema/CSApply.cpp (hunk 1 above)
Comment (DougGregor) [reply lib/Sema/CSApply.cpp]: Ah, it's incomplete. It's where the next bit of `typeCheckBinding` will be moving.

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please test source compatibility

**Comment by DougGregor** [comment]
@swift-ci please test platform Linux

**Comment by DougGregor** [comment]
@swift-ci please test platform Linux

**Comment by DougGregor** [comment]
@swift-ci please test platform Linux

**Comment by DougGregor** [comment]
@swift-ci please smoke test Linux

**Pull request merged**